packages/shared/lib/services/proxy/request.ts

```diff
@@ -0,0 +1,189 @@
+import { getProxyRetryFromErr } from './retry.js';
+
+/**
+ * This simple class is responsible to execute a call to a third-party
+ * It's handling the logging and retry, nothing else.
+ *
+ * It's the same code for the SDK and for the Proxy API.
+ */
+export class ProxyRequest {
+    logger: (msg: MessageRowInsert) => MaybePromise<void>;
+    config: ApplicationConstructedProxyConfiguration;
+
+    constructor({ logger, proxyConfig }: { logger: (msg: MessageRowInsert) => MaybePromise<void>; proxyConfig: ApplicationConstructedProxyConfiguration }) {
+        this.logger = logger;
+        this.config = proxyConfig;
+    }
+
+    /**
+     * Prepare the request and execute.
+     * Most code should use this method.
+     */
+    public async call(): Promise<Result<AxiosResponse>> {
+        const options: AxiosRequestConfig = {};
+
+        if (this.config.responseType) {
+            options.responseType = this.config.responseType;
+        }
+
+        if (this.config.data) {
+            options.data = this.config.data;
+        }
+
+        const { method } = this.config;
+
+        options.url = buildProxyURL(this.config);
+        options.method = method;
+
+        options.headers = buildProxyHeaders(this.config, options.url);
+
+        return await this.request({ axiosConfig: options });
+    }
+
+    /**
+     * Send a request to the third-party with retry.
+     * Only use this if you want a different axiosConfig than the one created in call()
+     */
+    public async request({ axiosConfig }: { axiosConfig: AxiosRequestConfig }): Promise<Result<AxiosResponse>> {
+        try {
+            const response = await retryFlexible<Promise<AxiosResponse>>(
+                async (retryAttempt) => {
+                    const start = new Date();
+                    try {
+                        const res = await axios.request(axiosConfig);
+                        await this.logResponse({ response: res, axiosConfig, retryAttempt, start });
+                        return res;
+                    } catch (err) {
+                        await this.logErrorResponse({ error: err, axiosConfig, retryAttempt, start });
+                        throw err;
+                    }
+                },
+                {
+                    max: this.config.retries || 0,
+                    onError: async ({ err, nextWait, max, attempt }) => {
+                        const retry = getProxyRetryFromErr({ err, proxyConfig: this.config });
+                        if (retry.retry) {
+                            await this.logger({
+                                type: 'log',
+                                level: 'warn',
+                                source: 'internal',
+                                createdAt: new Date().toISOString(),
+                                message: `Retrying HTTP call (reason: ${retry.reason}). Waiting for ${retry.wait ? retry.wait : nextWait}ms [${attempt}/${max}]`
```

Comment:
Would be nice to know what status code triggered it

Reply:
I realize that it can be seen in the error log right before it but I still think it would be good to be very explicit if possible.

Reply:
yeah agreed 👍🏻

Reply:
fixed in [122eee1](https://github.com/NangoHQ/nango/pull/3540/commits/122eee1535c95a4f9668b27e8e0a79019aef81fb)